SVG <stop> offset attribute incorrectly accepts invalid values with trailing garbage

Ahmad-S792 opened the pull request:
#### 808edf62942c254bd4fab37f78369061fa53d320
<pre>
SVG &lt;stop&gt; offset attribute incorrectly accepts invalid values with trailing garbage
<a href="https://bugs.webkit.org/show_bug.cgi?id=304794">https://bugs.webkit.org/show_bug.cgi?id=304794</a>
<a href="https://rdar.apple.com/167356988">rdar://167356988</a>

Reviewed by Nikolas Zimmermann.

The offset attribute parser was using String::toFloat() which silently ignores
trailing garbage. Per the SVG specification, attribute values with trailing
non-numeric characters should be treated as invalid and fall back to the
default value of 0.

Fixed by using parseNumber() from SVGParserUtilities which properly validates
the entire string. Also optimized to use StringView to avoid unnecessary string
copies.

* Source/WebCore/svg/SVGStopElement.cpp:
(WebCore::SVGStopElement::attributeChanged):
* LayoutTests/svg/parser/whitespace-number.html: Added.
* LayoutTests/svg/parser/whitespace-number-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/305036@main">https://commits.webkit.org/305036@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d81db35e9cec042360e1786a2fd5974f5f16bf4e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137236 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9596 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48523 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144987 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90209 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7cee0f49-b980-4b50-8c18-fa55b068e4fb) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/139108 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10299 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9723 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104953 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6b2c095f-4032-40cd-b471-885efc601ce1) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140181 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7601 "Found 1 new test failure: fast/forms/ios/select-option-removed-update.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122994 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85797 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/fa936b98-63e0-4eaf-8eaa-4146f5789fd3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7238 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4952 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5574 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116597 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41152 "Found 2 new test failures: http/tests/xmlhttprequest/cross-origin-cookie-storage.html imported/w3c/web-platform-tests/streams/piping/abort.any.worker.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147743 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9279 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41714 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113313 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9297 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7812 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113647 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28864 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7158 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119244 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63787 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9328 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37298 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9054 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72893 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9268 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9120 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nullhook commented on Source/WebCore/svg/SVGStopElement.cpp:
```diff
@@ -56,10 +56,19 @@ Ref<SVGStopElement> SVGStopElement::create(const QualifiedName& tagName, Documen
+        if (isPercentage)
+            valueView = valueView.left(valueView.length() - 1);
+
+        auto parsedValue = parseNumber(valueView, SuffixSkippingPolicy::DontSkip);
```
reordering sources could potentially break the usage of `parseNumber` here. 

please consider adding `#include "SVGParserUtilities.h"` in this file.

Ahmad-S792 replied:
Done. Added it.

nikolaszimmermann reviewed (approved):
Agreed thanks Ahmad!

Ahmad-S792 committed:
SVG <stop> offset attribute incorrectly accepts invalid values with trailing garbage
https://bugs.webkit.org/show_bug.cgi?id=304794
rdar://167356988

Reviewed by Nikolas Zimmermann.

The offset attribute parser was using String::toFloat() which silently ignores
trailing garbage. Per the SVG specification, attribute values with trailing
non-numeric characters should be treated as invalid and fall back to the
default value of 0.

Fixed by using parseNumber() from SVGParserUtilities which properly validates
the entire string. Also optimized to use StringView to avoid unnecessary string
copies.

* Source/WebCore/svg/SVGStopElement.cpp:
(WebCore::SVGStopElement::attributeChanged):
* LayoutTests/svg/parser/whitespace-number.html: Added.
* LayoutTests/svg/parser/whitespace-number-expected.txt: Added.

Canonical link: https://commits.webkit.org/305036@main
